[Index Management] Support data retention on Data Streams tab

alisonelizabeth opened the pull request:
This PR adds support for viewing the configured data retention value per data stream on the Data Streams tab in Index Management.

On **serverless**, a data stream can be managed by data stream lifecycle.

On **stateful**, a data stream can be managed by data stream lifecycle or ILM, or both. 

Fixes https://github.com/elastic/kibana/issues/165134
Fixes https://github.com/elastic/kibana/issues/152720
Fixes https://github.com/elastic/kibana/issues/154259
Fixes https://github.com/elastic/kibana/issues/165143

Note: Skipping release note for now, however, we will want to add one once the remaining work is complete (meta: https://github.com/elastic/kibana/issues/154256)

### How to test
1. Create a data stream managed via data stream lifecycle ([docs](https://www.elastic.co/guide/en/elasticsearch/reference/master/tutorial-manage-new-data-stream.html)). The configured data retention should appear on the list view and details flyout.
2. [Stateful only] Create a data stream managed via ILM ([docs](https://www.elastic.co/guide/en/elasticsearch/reference/current/set-up-a-data-stream.html)). There should be no data retention value. On the details flyout, there should be a link to the ILM policy.
3. [Stateful only] Create a data stream managed by ILM and data stream lifecycle. The configured data retention should appear on the list view and details flyout. The ILM policy name should also appear on the details flyout.

### Screenshots
<img width="1212" alt="Screenshot 2023-09-06 at 3 49 54 PM" src="https://github.com/elastic/kibana/assets/5226211/b51ef882-6787-4902-8b03-a7a2cf93f208">

<img width="274" alt="Screenshot 2023-09-06 at 12 52 03 PM" src="https://github.com/elastic/kibana/assets/5226211/e140dd11-7528-4051-a6eb-70e58f7fcd25">

<img width="509" alt="Screenshot 2023-09-06 at 12 52 10 PM" src="https://github.com/elastic/kibana/assets/5226211/a9cccd3c-5556-4245-8591-cfbf0d2a6404">

<img width="268" alt="Screenshot 2023-09-06 at 12 52 15 PM" src="https://github.com/elastic/kibana/assets/5226211/20dcbd39-3ab1-434c-8e60-f6ea476d3ea6">

jacksonbroshear commented:
Unsubscribe

On Tue, Sep 5, 2023, 1:53 PM kibana-ci ***@***.***> wrote:

> ⏳ Build in-progress, with failures
>
>    - Buildkite Build
>    <https://buildkite.com/elastic/kibana-pull-request/builds/156615>
>    - Commit: c41885c
>    <https://github.com/elastic/kibana/commit/c41885c1006e324a7abd626b80a8c22e783eaf3b>
>    - Interpreting CI Failures
>    <https://www.elastic.co/guide/en/kibana/current/interpreting-ci-failures.html>
>
> Failed CI Steps
>
>    - Checks
>    <https://buildkite.com/elastic/kibana-pull-request/builds/156615#018a6711-dab6-4b4a-8d77-09c612838fa1>
>
> History
>
>    - 💔 Build #156603
>    <https://buildkite.com/elastic/kibana-pull-request/builds/156603>
>    failed 79cb0ab
>    <https://github.com/elastic/kibana/commit/79cb0ab478cbc0ed58464bf0781f43ae4680e58d>
>    - 💔 Build #154982
>    <https://buildkite.com/elastic/kibana-pull-request/builds/154982>
>    failed 1cd71e3
>    <https://github.com/elastic/kibana/commit/1cd71e32dea9aa015c5a64cc0954fe48944ae9d6>
>
> To update your PR or re-run it, just comment with:
> @elasticmachine merge upstream
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/kibana/pull/165263#issuecomment-1707296757>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AABICVKDVHHT4LXRARKUTX3XY6GONANCNFSM6AAAAAA4E6IGPI>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

alisonelizabeth commented on x-pack/plugins/index_management/public/application/sections/home/data_stream_list/data_stream_table/data_stream_table.tsx:
```diff
@@ -71,7 +71,6 @@ export const DataStreamTable: React.FunctionComponent<Props> = ({
     render: (health: DataStream['health']) => {
       return <DataHealth health={health} />;
     },
-    width: '100px',
```
This was causing some weird spacing issues in the table.

alisonelizabeth commented on x-pack/plugins/index_management/common/types/data_streams.ts:
```diff
@@ -5,20 +5,20 @@
 
 type TimestampField = TimestampFieldFromEs;
 
-interface MetaFromEs {
```
I refactored some of the server-side code to leverage the types from `@elastic/elasticsearch` rather than our own custom interfaces

alisonelizabeth commented:
@elasticmachine merge upstream

yuliacech reviewed (approved):
Thanks a lot for working on data retention, @alisonelizabeth! 
Tested locally and the information is displayed as expected, code changes LGTM too 👍

gchaps reviewed (commented):
Index appears twice in the flyout. 

<img width="335" alt="Screenshot 2023-09-12 at 2 00 51 PM" src="https://github.com/elastic/kibana/assets/33642766/ac0c5345-64f0-452b-8dd8-b19f47028fa1">

alisonelizabeth commented on x-pack/plugins/index_management/public/application/sections/home/data_stream_list/data_stream_detail_panel/data_stream_detail_panel.tsx:
```diff
@@ -195,6 +247,7 @@ export const DataStreamDetailPanel: React.FunctionComponent<Props> = ({
           defaultMessage: 'Cumulative count of backing indices created for the data stream',
```
Good question. This was pre-existing and comes directly from the ES API: https://www.elastic.co/guide/en/elasticsearch/reference/current/data-streams.html#data-streams-generation

alisonelizabeth commented:
@gchaps I've addressed most of your feedback in https://github.com/elastic/kibana/pull/165263/commits/f48949bffa6db648b38d4acd9d184ceab9596c2d. Let me know what you think about the final copy for the tooltip in the table. Here's the latest screenshot.

<img width="331" alt="Screenshot 2023-09-14 at 11 41 11 AM" src="https://github.com/elastic/kibana/assets/5226211/3959a59d-8e63-4981-8b82-4419a7329ac1">

tylerperk commented:
> Let me know what you think about the final copy for the tooltip in the table. Here's the latest screenshot.
> 
> <img alt="Screenshot 2023-09-14 at 11 41 11 AM" width="331" src="https://user-images.githubusercontent.com/5226211/268033013-3959a59d-8e63-4981-8b82-4419a7329ac1.png">

@alisonelizabeth Something that comes up is how well ILM and now DSL enable timely deletion for compliance reasons (GDPR etc). We don't delete the data exactly after 7 days (in the above example) so we should be careful to not give that impression. I wonder if we could improve the first sentence a little as the first time I read it I felt a bit like it implies a guarantee which we know won't be satisfied. It is more like "the minimum amount of time data will be available before it is automatically deleted" or "data will be be kept at least this long before it is automatically deleted" or ?

gchaps commented:
Here's a suggestion based on what @tylerperk's comment:

>Data will be be kept at least this long before it is automatically deleted. Only applies to data streams managed by a data stream lifecycle.  This value might not apply to all data if the data stream also has an index lifecycle policy.

alisonelizabeth commented:
Thanks @gchaps! I updated the text in https://github.com/elastic/kibana/pull/165263/commits/264952a0709f3975ab975b1ef0010d5163633c6c.
